dev/test/pool.ts

```diff
@@ -327,4 +327,29 @@ describe('Client pool', () => {
         expect(err).to.equal('The client has already been terminated');
       });
   });
+
+  it('waits for existing operations to complete before releasing clients', done => {
+    const clientPool = new ClientPool<{}>(1, 0, () => {
+      return {};
+    });
+    const deferred = new Deferred<void>();
+    let terminated = false;
+
+    // Run operation that completes after terminate() is called.
+    clientPool.run(REQUEST_TAG, () => {
+      return deferred.promise;
+    });
+    const terminateOp = clientPool.terminate().then(() => {
+      terminated = true;
+    });
+
+    setImmediate(async () => {
```

Comment:
Just glancing at this, I don't quite follow why we need `setImmediate`. Can you add a comment (or remove it if it isn't necessary).